docker/docker_image_dest.go

```diff
@@ -99,35 +108,50 @@ func (d *dockerImageDestination) PutBlob(digest string, stream io.Reader) error
 	logrus.Debugf("Uploading %s", uploadURL)
 	res, err = d.c.makeRequest("POST", uploadURL, nil, nil)
 	if err != nil {
-		return err
+		return "", -1, err
 	}
 	defer res.Body.Close()
 	if res.StatusCode != http.StatusAccepted {
 		logrus.Debugf("Error initiating layer upload, response %#v", *res)
-		return fmt.Errorf("Error initiating layer upload to %s, status %d", uploadURL, res.StatusCode)
+		return "", -1, fmt.Errorf("Error initiating layer upload to %s, status %d", uploadURL, res.StatusCode)
 	}
 	uploadLocation, err := res.Location()
 	if err != nil {
-		return fmt.Errorf("Error determining upload URL: %s", err.Error())
+		return "", -1, fmt.Errorf("Error determining upload URL: %s", err.Error())
 	}
 
 	// FIXME: DELETE uploadLocation on failure
 
+	h := sha256.New()
+	pr, pw := io.Pipe()
+	bufin := bufio.NewReader(stream)
+	mw := io.MultiWriter(h, pw)
+	go func() {
+		bufin.WriteTo(mw)
+		pw.Close()
+	}()
+	hash := h.Sum(nil)
```

Comment:
Well, there is https://github.com/docker/distribution/blob/master/registry/client/blob_writer.go . Either we can rely on the `ReadFrom` in there forcing servers to support that kind of request, or we can do what `Write` is doing and chunk the input into buffers of a known (small?) size. Doing the latter might require some tuning so that we aren’t slowed down by constant roundtrips, though.